scripts/generate-unicode-id-parts.js

```diff
@@ -0,0 +1,158 @@
+    if (this.lastTrailSurrogate !== this.firstTrailSurrogate) {
+      text += '-\\u' + toHex4Digits(this.lastTrailSurrogate);
+    }
+    return text;
+  }
+}
+
+class Writer {
+  constructor() {
+    /** @type {number} */
+    this.lastCodePoint = -1;
+    /** @type {NonSurrogateRange[]} */
+    this.nonSurrogateRanges = [];
+    /** @type {LeadSurrogateRange[]} */
+    this.surrogateRanges = [];
+    /** @type {NonSurrogateRange} */
+    this.nonSurrogateRange;
+    /** @type {LeadSurrogateRange} */
+    this.leadSurrogateRange;
+    /** @type {TrailSurrogateRange} */
+    this.trailSurrogateRange;
+  }
+
+  /** @param {number} codePoint */
+  push(codePoint) {
+    if (codePoint <= this.lastCodePoint)
+      throw new Error('Code points must be added in order.');
+    this.lastCodePoint = codePoint;
+
+    if (codePoint < MAX_UNICODE_NON_SURROGATE) {
+      if (
+        this.nonSurrogateRange &&
+        this.nonSurrogateRange.lastCodePoint === codePoint - 1
+      ) {
+        this.nonSurrogateRange.lastCodePoint = codePoint;
+        return;
+      }
+      this.nonSurrogateRange = new NonSurrogateRange(codePoint);
+      this.nonSurrogateRanges.push(this.nonSurrogateRange);
+    } else {
+      const leadSurrogate = Math.floor((codePoint - 0x10000) / 0x400) + 0xd800;
+      const trailSurrogate = ((codePoint - 0x10000) % 0x400) + 0xdc00;
+      if (
+        !this.leadSurrogateRange ||
+        this.leadSurrogateRange.leadSurrogate !== leadSurrogate
+      ) {
+        this.trailSurrogateRange = undefined;
+        this.leadSurrogateRange = new LeadSurrogateRange(leadSurrogate);
+        this.surrogateRanges.push(this.leadSurrogateRange);
+      }
+
+      if (
+        this.trailSurrogateRange &&
+        this.trailSurrogateRange.lastTrailSurrogate === trailSurrogate - 1
+      ) {
+        this.trailSurrogateRange.lastTrailSurrogate = trailSurrogate;
+        return;
+      }
+
+      this.trailSurrogateRange = new TrailSurrogateRange(trailSurrogate);
+      this.leadSurrogateRange.ranges.push(this.trailSurrogateRange);
+    }
+  }
+
+  toString() {
+    let first = this.nonSurrogateRanges.join('');
+    let second = this.surrogateRanges.join('|');
+    return first && second
+      ? `([${first}]|${second})`
+      : first
+      ? `[${first}]`
+      : second
+      ? `(${second})`
+      : '';
+  }
+}
+
+const MAX_UNICODE_NON_SURROGATE = 0xffff;
+const MAX_UNICODE_CODEPOINT = 0x10ffff;
+const isStart = c => /\p{ID_Start}/u.test(c);
+const isContinue = c => /\p{ID_Continue}/u.test(c);
+
+let idStartWriter = new Writer();
+let idContinueWriter = new Writer();
+
+for (let cp = 0; cp <= MAX_UNICODE_CODEPOINT; cp++) {
+  const ch = String.fromCodePoint(cp);
+  if (isStart(ch)) {
+    idStartWriter.push(cp);
+  }
+  if (isContinue(ch)) {
+    idContinueWriter.push(cp);
+  }
+}
+
+console.log(`/**
+* Generated by scripts/generate-unicode-id-parts.js on node ${
+  process.version
+} with unicode ${process.versions.unicode}
```

Comment:
So this depends on a Node version, and might be flaky/inconsistent? I'm wondering if it's possible just do a predictable subset, and update it periodically (even if manually).

Also, where/when will we run this script? Is it a part of the build script?

Reply:
It was run manually. Its unlikely this will change often, usually only if you choose to target a new version of NodeJS that itself includes a new version of the Unicode standard.

Reply:
OK, we can add it to the build script as one of the steps at some point.

Reply:
I'll copy a comment I made on the TypeScript implementation here: https://github.com/microsoft/TypeScript/pull/32096#issuecomment-524203481

The build script could be simplified and made to not rely on the current Node.js/V8 + Unicode version:

```js
const ID_Start = require('unicode-12.1.0/Binary_Property/ID_Start/code-points.js');
const ID_Continue = require('unicode-12.1.0/Binary_Property/ID_Continue/code-points.js');

// ...then add the other needed characters, and write the two arrays to disk.
```

Links to how other tooling implements this:

- Esprima: https://github.com/jquery/esprima/blob/master/tools/generate-identifier-regex.js
- @babel/parser: https://github.com/babel/babel/blob/master/packages/babel-parser/scripts/generate-identifier-regex.js
- Acorn: https://github.com/acornjs/acorn/blob/master/bin/generate-identifier-regex.js
- JSHint: https://github.com/jshint/jshint/blob/master/scripts/generate-identifier-data.js
- XRegExp: https://github.com/slevithan/xregexp/blob/master/tools/scripts/utils.js